nav2_route/src/breadth_first_search.cpp

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2020 Samsung Research America
+// Copyright (c) 2023 Joshua Wallace
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_route/breadth_first_search.hpp"
+
+#include <iostream>
+#include <queue>
+#include <unordered_map>
+
+namespace nav2_route
+{
+
+void BreadthFirstSearch::setCostmap(nav2_costmap_2d::Costmap2D * costmap)
+{
+  std::cout << "Set costmap " << std::endl;
+  costmap_ = costmap;
+
+  int x_size = static_cast<int>(costmap_->getSizeInCellsX());
+  int y_size = static_cast<int>(costmap_->getSizeInCellsY());
+
+  max_index_ = x_size * y_size;
+  x_size_ = x_size;
+
+  neighbors_grid_offsets_ = {-1, +1,
```

Comment:
I'd do some testing and make sure that 8 connected has a benefit over 4 connected. I assume for a BFS it does, but I wouldn't assume that without testing

Reply:
ex https://medium.com/analytics-vidhya/part-i-breadth-first-search-using-grid-dc41a5f41663 you can see that using a 4 connected neighborhood that the expansion waves aren't equal, so one might be explored first that's axially aligned with the X/Y axis which is actually up to sqrt(2) further away from one that's closer on a diagonal.

Reply:
Do you have an tips for the visualization? Publish out a occgrid?

Reply:
Depends on what strikes your fancy, but probably an occ grid. Maybe take a snapshot every so often of the number of expanded cells and color them black to see how they expand over time. Or, do them all at once and visualize the full grid color-coded by visitation time so you can see the pattern emerging.